Workiva/over_react
# CPLAT-9203 CPLAT-9204 Implement new mixin-based boilerplate, refactor builder parsing/codegen

# Motivation
Support [the new mixin-based boilerplate](https://github.com/Workiva/over_react/blob/new_boilerplate_wip/doc/new_boilerplate_migration.md).

# Changes
#### High-level changes:
- Refactor boilerplate parsing so it can support both the new and old boilerplate side-by-side, without the parsing of the two conflicting with each other.
    - Bonus: multiple components per file are now supported in all boilerplate versions.
- Refactor boilerplate codegen to be more modular, to more easily support the new boilerplate.
- Add support for new boilerplate in builder, add tests.
- Update builder examples to use new boilerplate.

#### Changes not included, which will come in later PRs:
- Update web/app dirs to use new boilerplate
    - Done alongside this work, but not included to reduce diff. See PR: https://github.com/Workiva/over_react/pull/482
- Update builder README (will probably contain a lot of the same info as this PR description)
- Implement function components

## Refactor boilerplate parsing
#### Before, parsing looked like this:
_All in declaration_parsing.dart:_
- Parse a file, and find all nodes with matching over_react annotations
- Check whether the count of Factory/Props/State/Component annotations is correct to declare either 0 or 1 component. If not, emit an error.
- For props/state classes/mixins, determine whether they're part of the backwards-compatible or Dart-2-only boilerplate
- Perform validation on all nodes
- Emit `ParsedDeclarations`, containing:
    - (optional) nodes comprising a single component declaration
    - all nodes relating to declarations of props/state mixins and abstract props/state classes

#### Now, parsing looks like this:
_In members_from_ast.dart:_
- Parse a file, and in `detectBoilerplateMembers` find all nodes that could potentially be part of over_react_boilerplate
    - Whether a node could be a specific kind of boilerplate is determined by a number of factors including annotations, structure, naming, and supertypes
    - For each node that could be boilerplate, determine a "version confidence score", which represents how likely they are to be a given boilerplate version. The versions:
        - `Version.v2_backwardsCompat` - props/state have stubbed companion classes
        - `Version.v3_dart2Only` - no companion classes, the public props/state are generated
        - `Version.v4_mixinBased` - the new boilerplate!
    - If a score is not sufficiently high for any version, it won't get generated, and this is how we avoid false positives. We'll talk about that more later.
    - Instantiate and emit a subclass of `BoilerplateMember` containing this node and version

_In declarations_from_members.dart:_
- Group the `BoilerplateMember` into `BoilerplateDeclarations` within `getBoilerplateDeclarations`:
    - Group them together into complete boilerplate declarations, like class components, mixins, etc.
        - Make abstract props/state and props/state mixins into declarations, since they're single members that don't need grouping
        - Loop through the factories, which are the entrypoint for all remaining declaration types, and find the props/state classes and components with matching names or type parameters
            - Special case: if there's a single legacy component declared in a file, group it together regardless of the names to provide backwards compatibility with the old builder.
    - For each potential declaration, resolve the boilerplate version, which is the result of averaging the confidence scores of all constituent members and picking the version with the highest score
        - This allows:
            - members that look the same across different boilerplate versions (e.g., factories) and thus have equal confidences for each version to have less influence over the end version than other members
            - members that aren't declared quite correctly due to user error to still be grouped with the intended other members, and validated against the same version.
        - If the score does not exceed a threshold, then we can't confidently say that this is a true over_react declaration that needs code generation, and we skip it
        - If the score does exceed a threshold, instantiate and emit a subclass of `BoilerplateDeclaration` with the members and resolved version
    - Loop through mismatched declarations that were not grouped and provide error messages for those whose confidence is above the threshold.
- Validate the declarations, which includes validating child members against the resolved version.

## Refactor boilerplate codegen
#### Before, codegen looked like this:
_In impl_generation.dart:_
- For each file, a `ParsedDeclarations` instance is passed to `ImplGenerator.generate`, which:
    - Generates code for one or more component
    - Generates code for all given props/state mixins and abstract props/state classes
- Code generation logic for the different pieces is split among different helper methods, which are passed the data they need as arguments

#### Now, codegen looks like this:
_In impl_generation.dart:_
- For each file, we iterate over aa list of `BoilerplateDeclarations` and pass each to `ImplGenerator.generate`, which delegates generation logic to a number of newly-instantiated generator subclasses which generate for each part of the boilerplate:
    - component class
    - component factory proxy
    - props impl + uiFactory, state impl
    - accessors mixins

_In codegen/*\_generator.dart:_
- Each generator subclass takes in the data it needs as constructor args and stores it as fields
    - These are available to all helper methods, reducing the need for passing around so many arguments
    - Name-related logic is split out into `names.dart`

Codegen logic remains the same for legacy boilerplate, just rearranged and shared with codegen for new boilerplate.

__There should be no changes to generated files for old boilerplate.__

# Review
_[See CONTRIBUTING.md][contributing-review-types] for more details on review types (+1 / QA +1 / +10) and code review process._

  

FYI @corwinsheahan-wf @evanweible-wf @kealjones-wk 

## QA Checklist
        - Ensure example/builder/ app runs correctly
        - Ensure app/web work as expected in follow-up PR: https://github.com/Workiva/over_react/pull/482
        - Run new builder on wdesk and ensure there are no changes to generated output (talk to Greg/Joe for script to help with this)


[contributing-review-types]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#review-types
[contributing-manual-testing]: https://github.com/Workiva/over_react/blob/master/CONTRIBUTING.md#manual-testing-criteria
